Add chapter on the Bevy Organization to the Contributing Guide (#1596)
bevyengine/bevy-website: content/contributing/project-information/release-process.md

+++
title = "Release Process"
insert_anchor_links = "right"
[extra]
weight = 3
+++

> Bevy uses three-months-long development cycles, delimited by a weeks-long rolling release process.

Reviewer: ```suggestion
Bevy uses a three-months-long development cycles, delimited by a weeks-long rolling release process.
```

Author: I don't agree with this suggestion!

Reviewer: Now that I look at it again, I missed the s at the end of cycles. My bad.

> This means that bevy releases a new version roughly once every three to four months. In an effort to avoid crunch and encourage sustainable development, we do not pick specific dates or deadlines.

Reviewer: ```suggestion
This means that Bevy releases a new version roughly once every three to four months. In an effort to avoid crunch and encourage sustainable development, we do not pick specific dates or deadlines.
```
Bevy is a proper noun.

## Development Cycles

> During a three-month development cycle, maintainers collect important PRs and issues into a github Milestone. Issues included in the milestone are given priority, and the community generally makes a best-effort attempt to merge as many of these as possible over the duration of the cycle.

Reviewer 2: ```suggestion
During a three-month development cycle, maintainers collect important PRs and issues into a GitHub Milestone. Issues included in the milestone are given priority, and the community generally makes a best-effort attempt to merge as many of these as possible over the duration of the cycle.
```

> During a three-month development cycle, maintainers collect important PRs and issues into a GitHub Milestone. Issues included in the milestone are given priority, and the community generally makes a best-effort attempt to merge as many of these as possible over the duration of the cycle.

Reviewer: ```suggestion
During a three-month development cycle, Maintainers collect important PRs and issues into a GitHub Milestone. Issues included in the milestone are given priority, and the community generally makes a best-effort attempt to merge as many of these as possible over the duration of the cycle.
```

> In the lead-up to a release, generally about about two and a half months after the previous, we begin to aggressively purge the milestone of things that seem unlikely to be ready in time. The main things that tend to remain are:

Reviewer 2: ```suggestion
In the lead-up to a release (generally about about two and a half months after the previous) we begin to aggressively purge the milestone of things that seem unlikely to be ready in time. The main things that tend to remain are:
```
This works better as a parenthetical in my opinion.

- Quick and easy features under active development.
- Fixes for important regressions or crashes.
- Large or impressive features that only need minor polish.

> When the milestone is complete, we publish a release candidate (ex. `0.14.0-rc.1`) and begin the release process.

Reviewer 3: Also mention that a Working Group will be created for release prep, to help organize and coordinate the process.

## Release Candidates

> In the run-up to a release, Bevy publishes a new release candidate about once a week. Publishing release candidates allows users to identify bugs and regressions, gives ecosystem crates time to update so they can be ready closer to release day, and provides bevy contributors extra time to write release notes and migration guides.

Reviewer 3: I think this last bit isn't worded strongly enough. I frankly think it deserves its own paragraph.

Once again we are neglecting to acknowledge the monumental effort required for release prep (the blog post, etc) and how much time that takes!

The way I see it, the benefit of the RC period is just as much (if not more) about giving maintainers + the community time for release prep work (such as writing the blog post), as it is about code quality.

I think it should be more clearly communicated.

Release prep requires a monumental amount of effort and I believe that should be acknowledged. Y'all need the time to be able to do it stress-free and crunch-free.

Previous releases have been very close to driving people to burnout because this isn't more explicitly addressed.

There is a lot of focus on "we don't set specific deadlines to avoid crunch for devs" (in the previous section about the dev cycle), but not enough on avoiding crunch for the people involved in release prep!

I have been advocating for a clear official "release prep period" after the 3-month dev cycle primarily for this reason!

Author: I want to tackle this in a dedicated follow-up :)

> In the run-up to a release, Bevy publishes a new release candidate about once a week. Publishing release candidates allows users to identify bugs and regressions, gives ecosystem crates time to update so they can be ready closer to release day, and provides bevy contributors extra time to write release notes and migration guides.

Reviewer: ```suggestion
In the run-up to a release, Bevy publishes a new release candidate about once a week. Publishing release candidates allows users to identify bugs and regressions, gives ecosystem crates time to update so they can be ready closer to release day, and provides Bevy contributors extra time to write release notes and migration guides.
```

When publishing a release candidate, we also announce a general target release date. This is generally about one week out. If critical bugs or regressions are discovered during this testing window, we fix them and publish a new release candidate. Then we either

> - Announce a new date, if we believe more testing time is needed than the current window.

Reviewer 2: ```suggestion
- Announce a new date, if we believe more testing time is needed than the current window, or
```

> - Or keep the same target release date if we believe there is enough time in the current window.

Reviewer 2: ```suggestion
- Keep the same target release date if we believe there is enough time in the current window.
```

> - Keep the same target release date if we believe there is enough time in the current window.

Reviewer 3: Again, can we please not make the RC period all about testing and code quality. This section also needs a point about "we need more time for release prep".

Author: Ditto on follow-up :)

Reviewer 4: I was just trying to distill Cart's original description, as he posted it in discord. Totally agree, and let's bikeshed in follow-up.

When we successfully hit a release candidate's target date, we do the final release.

> ## Release Checklist

Reviewer 3: I am not reviewing this, because I don't know much about it.

> When making a release, the maintainers follow these checklists:

Reviewer: ```suggestion
When making a release, the Maintainers follow these checklists:
```
Going by references in the other documents, this should be capitalized as a proper noun, perhaps.

### Minor Version

#### Minor Pre-release

1. Check regressions tag.

> 2. Check appropriate milestone, and close it.

Reviewer 2: ```suggestion
2. Check appropriate milestone and close it.
```

3. Check GitHub Projects page for staleness.
4. Update change log.
5. Create migration guide.
6. Write blog post.
7. Update book.
8. Bump version number for all crates, using the "Release" workflow.
   1. Change the commit message to be nicer.
9. Create tag on GitHub.

> 10. Edit Github Release. Add links to the `Release announcement` and `Migration Guide`.

Reviewer 2: ```suggestion
10. Edit GitHub Release. Add links to the `Release announcement` and `Migration Guide`.
```

11. Bump `latest` tag to most recent release.

> 12. Run [the `update-screenshots` workflow] to update screenshots. _This will block blog post releases (and take ~40 minutes) so do it early_.

Reviewer: ```suggestion
12. Run [the `update-screenshots` workflow] to update screenshots. *This will block blog post releases (and take ~40 minutes) so do it early*.
```
Other documents use asterisks for italics; we should maintain that consistency.

> 13. Run [the `build-wasm-examples` workflow] to update wasm examples.

Reviewer: ```suggestion
13. Run [the `build-wasm-examples` workflow] to update Wasm examples.
```
Wasm is capitalized as an abbreviation (not acronym).

#### Minor Release

1. Release on crates.io using `bash tools/publish.sh`
2. Announce on:
    1. HackerNews
    2. Twitter
    3. Reddit: /r/bevy, /r/rust, /r/rust_gamedev
    4. Discord: Bevy, Game Development in Rust, Rust Programming Language Community
    5. This Month in Rust Game Development newsletter
    6. This Week in Rust newsletter

#### Minor Post-release

1. Bump version number for all crates to next versions, as `0.X-dev`, using the "Post-release version bump" workflow, to ensure properly displayed version for [Dev Docs](https://dev-docs.bevyengine.org/bevy/index.html).

> 2. Update Bevy version used for Bevy book code validation to latest release.

Reviewer: ```suggestion
2. Update Bevy version used for the Bevy website's `learning-code-examples` tool (code example validation and formatting for the learning materials) to latest release.
```
Clarifying this.

### Patch

#### Patch Pre-release

1. Check appropriate milestone.
2. Close the milestone, open the next one if anything remains and transfer them.
3. Bump version number for all crates, using [the command from the "Release" workflow] locally, with `patch` for the new version.

>     - Change the commit message to be nicer

Reviewer 2: ```suggestion
    - Change the commit message to be nicer.
```

4. Create tag on GitHub.

> 5. Edit Github Release. Add link to the comparison between this patch and the previous version.

Reviewer 2: ```suggestion
5. Edit GitHub Release. Add link to the comparison between this patch and the previous version.
```

6. Bump `latest` tag to most recent release.
7. Run the [`update-screenshots` workflow] to update screenshots.

> 8. Run this [`build-wasm-examples` workflow] to update wasm examples.

Reviewer: ```suggestion
8. Run this [`build-wasm-examples` workflow] to update Wasm examples.
```

#### Patch Release

1. Release on crates.io
    - `bash tools/publish.sh`
2. Announce on:
    1. Discord: Bevy

#### Patch Post-Release

### Release Candidate

#### RC Pre-Release

1. Check appropriate milestone.
2. Create a branch for the release.
3. Bump version number for all crates, using [the command from the "Release" workflow] locally, with `rc` for the new version.

>     - Change the commit message to be nicer

Reviewer 2: ```suggestion
    - Change the commit message to be nicer.
```

4. Create tag on GitHub.

> 5. Edit Github Release. Add link to the comparison between this rc and the previous version.

Reviewer 2: ```suggestion
5. Edit GitHub Release. Add link to the comparison between this rc and the previous version.
```

#### RC Release

1. Release on crates.io
    - `bash tools/publish.sh`
2. Announce on:
    1. Discord: Bevy, #dev-announcements

#### RC Post-Release

> 1. Update Bevy version used for Bevy book code validation to latest release.

Reviewer: ```suggestion
1. UUpdate Bevy version used for the Bevy website's `learning-code-examples` tool (code example validation and formatting for the learning materials) to latest release.
```
Clarification ditto.

Author: ```suggestion
1. Update Bevy version used for the Bevy website's `learning-code-examples` tool (code example validation and formatting for the learning materials) to latest release.
Commit suggestion
Add suggestion to batch```
